### PR TITLE
[swiftmailer] Document whitelist option to email redirect

### PR DESCRIPTION
**Pull request by TerjeBr**
The redirectPlugin in switfmailer allows you to specify a white-list of regex's that matches email addresses that are still sent to, when redirecting all email to one address.

Code was added in PR [switfmailer/SwiftmailerBundle#19](https://github.com/symfony/SwiftmailerBundle/pull/19)

### CODE REVIEW

**Comment by TerjeBr** [comment]
Documntation of code added in [switfmailer/SwiftmailerBundle#19](https://github.com/symfony/SwiftmailerBundle/pull/19)

File: cookbook/email/dev_environment.rst [hunk 1]
```diff
@@ -119,6 +119,51 @@ the replaced address, so you can still see who it would have been sent to.
     These are ``X-Swift-Cc`` and ``X-Swift-Bcc`` for the ``CC`` and ``BCC``
     addresses respectively.
 
+Sending to a Specified Address, but with exceptions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Suppose you want to have all email sent to a specific address, instead
+of the address actually specified when sending the message (like in the above scenario),
+but you want certain email addresses not to be redirected in this way.
```
Comment (WouterJ) [inline cookbook/email/dev_environment.rst]: This has become a very long sentence, let's try to make it a bit shorter. What do you think about?

````
Suppose you want all emails to be sent to ``dev@example.com``, except
from the emails sent to ``admin@specialdomain.com``. In this case, you
can use the ``delivery_whitelist`` option:

.. configuration-block::

    .. code-block:: yaml

        # app/config/config_dev.yml
        swiftmailer:
            delivery_address: dev@example.com
            delivery_whitelist:
                # all email addresses matching this regex will *not* be
                # redirected to dev@example.com
                - "/@mydomain.com$/"

                # all emails sent to admin@specialdomain.com won't
                # be redirected to dev@example.com too
                - "/^admin@specialdomain.com$/"

    .. code-block:: xml

        <?xml version="1.0" charset="UTF-8" ?>
        <container xmlns="http://symfony.com/schema/dic/services"
            xmlns:swiftmailer="http://symfony.com/schema/dic/swiftmailer"
        >
            <swiftmailer:config delivery-address="dev@example.com">
                <!-- all email adress matching this regex will *not* be redirected
                    to dev@example.com -->
                <swiftmailer:delivery-whitelist-pattern>/@mydomain.com$/</swiftmaler:delivery-whitelist-pattern>

                <!-- all emails sent to admin@specialdomain.com won't be
                    redirectd to dev@example.com too -->
                <swiftmailer:delivery-whitelist-pattern>/^admin@specialdomain.com$/</swiftmailer:delivery-whitelist-pattern>
            </swiftmailer:config>
````

File: cookbook/email/dev_environment.rst [hunk 2]
```diff
@@ -119,6 +119,57 @@ the replaced address, so you can still see who it would have been sent to.
...
     addresses respectively.
 
+Sending to a Specified Address, but with exceptions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Suppose you normally in your dev environment want to have all email redirected to a specific address,
+(like in the above scenario to ``dev@example,com``).
+But then you may want email sent to some specific email addresses to go through after all,
+and not be redirected (even if it is in the dev environment).
+This can be done by adding the ``delivery_whitelist`` option:
```
Comment (WouterJ) [inline cookbook/email/dev_environment.rst]: could you please reformat this to only add line breaks after the first word that crosses the 72th character? I.e:

````rst
Suppose you normally in your dev environment want to have all email redirected
to a specific address, (like in the above scenario to ``dev@example,com``).
But then you may want email sent to some specific email addresses to go through
after all, and not be redirected (even if it is in the dev environment).  This
can be done by adding the ``delivery_whitelist`` option:
````

File: cookbook/email/dev_environment.rst [hunk 3]
```diff
@@ -119,6 +119,57 @@ the replaced address, so you can still see who it would have been sent to.
...
+        $container->loadFromExtension('swiftmailer', array(
+            'delivery_address'  => "dev@example.com",
+            'delivery_whitelist' => array(
+                '/@mydomain.com$/',
+                '/^admin@specialdomain.com$/'
+            ),
+        ));
+
+In the above example all mail will be redirected to ``dev@example.com``, except that mail to the single
+address ``admin@specialdomain.com`` and all mail to the domain ``mydomain.com`` will be delivered as normal.
```
Comment (WouterJ) [inline cookbook/email/dev_environment.rst]: same here

File: cookbook/email/dev_environment.rst (hunk 2 above)
Comment (TerjeBr) [inline cookbook/email/dev_environment.rst]: done

File: cookbook/email/dev_environment.rst (hunk 3 above)
Comment (TerjeBr) [inline cookbook/email/dev_environment.rst]: ok, done.

File: cookbook/email/dev_environment.rst [hunk 4]
```diff
@@ -119,6 +119,58 @@ the replaced address, so you can still see who it would have been sent to.
     These are ``X-Swift-Cc`` and ``X-Swift-Bcc`` for the ``CC`` and ``BCC``
     addresses respectively.
 
+Sending to a Specified Address, but with exceptions
```
Comment (javiereguiluz) [inline cookbook/email/dev_environment.rst]: In the Symfony documentation, the titles follow some special notation. That's why this should be reworded as follows (unless @xabbuh suggest other rewording):

```
Original: Sending to a Specified Address, but with exceptions

New: Sending to a Specified Address but with Exceptions
```

File: cookbook/email/dev_environment.rst [hunk 5]
```diff
@@ -119,6 +119,58 @@ the replaced address, so you can still see who it would have been sent to.
...
+
+        <!--
+            xmlns:swiftmailer="http://symfony.com/schema/dic/swiftmailer"
+            http://symfony.com/schema/dic/swiftmailer http://symfony.com/schema/dic/swiftmailer/swiftmailer-1.0.xsd
+        -->
+
+        <swiftmailer:config delivery-address="dev@example.com" />
+        <swiftmailer:delivery-whitelist>/@mydomain.com$/</swiftmailer:delivery-whitelist>
+        <swiftmailer:delivery-whitelist>/^admin@specialdomain.com$/</swiftmailer:delivery-whitelist>
+    .. code-block:: php
```
Comment (javiereguiluz) [inline cookbook/email/dev_environment.rst]: There is a missing blank line above this line

File: cookbook/email/dev_environment.rst [hunk 6]
```diff
@@ -119,6 +119,58 @@ the replaced address, so you can still see who it would have been sent to.
...
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # app/config/config_dev.yml
+        swiftmailer:
+            delivery_address: dev@example.com
+            delivery_whitelist:
+               # all email addresses matching this regex will *not* be
```
Comment (javiereguiluz) [inline cookbook/email/dev_environment.rst]: I don't know which is the common practice in the Symfony docs (maybe @WouterJ and @xabbuh can help us with this) but it's strange to add explanation comments in the PHP format but not include them in the XML and PHP formats. We have two solutions:

  * Repeat the same explanation for all formats.
  * Don't add these comments and just explain it using the regular text content.

File: cookbook/email/dev_environment.rst [hunk 7]
```diff
@@ -119,6 +119,58 @@ the replaced address, so you can still see who it would have been sent to.
...
+        // app/config/config_dev.php
+        $container->loadFromExtension('swiftmailer', array(
+            'delivery_address'  => "dev@example.com",
+            'delivery_whitelist' => array(
+                '/@mydomain.com$/',
+                '/^admin@specialdomain.com$/'
+            ),
+        ));
+
+In the above example all mail will be redirected to ``dev@example.com``,
```
Comment (javiereguiluz) [inline cookbook/email/dev_environment.rst]: I would reword this paragraph a bit, mostly because `mail` in English refers exclusively to the "postal  mail" and here we're always talking about email. Here it is a proposal:

```
In the above example all email messages will be redirected to ``dev@example.com``,
except messages sent to the ``admin@specialdomain.com`` address or to any email
address belonging to the domain ``mydomain.com``, which will be delivered as normal.
```

File: cookbook/email/dev_environment.rst [hunk 8]
```diff
@@ -119,6 +119,58 @@ the replaced address, so you can still see who it would have been sent to.
     These are ``X-Swift-Cc`` and ``X-Swift-Bcc`` for the ``CC`` and ``BCC``
     addresses respectively.
 
+Sending to a Specified Address, but with exceptions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Suppose you normally in your dev environment want to have all email redirected
+to a specific address, (like in the above scenario to ``dev@example,com``).
```
Comment (javiereguiluz) [inline cookbook/email/dev_environment.rst]: minor typo: `dev@example,com` -> `dev@example.com`

File: cookbook/email/dev_environment.rst (hunk 6 above)
Comment (WouterJ) [inline cookbook/email/dev_environment.rst]: We usually add the comments to all formats.

File: cookbook/email/dev_environment.rst [hunk 9]
```diff
@@ -119,6 +119,58 @@ the replaced address, so you can still see who it would have been sent to.
     These are ``X-Swift-Cc`` and ``X-Swift-Bcc`` for the ``CC`` and ``BCC``
     addresses respectively.
 
+Sending to a Specified Address, but with exceptions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Suppose you normally in your dev environment want to have all email redirected
```
Comment (cordoval) [inline cookbook/email/dev_environment.rst]: remove normally here

File: cookbook/email/dev_environment.rst [hunk 10]
```diff
@@ -119,6 +119,58 @@ the replaced address, so you can still see who it would have been sent to.
...
+to a specific address, (like in the above scenario to ``dev@example,com``).
+But then you may want email sent to some specific email addresses to go through
+after all, and not be redirected (even if it is in the dev environment).
```
Comment (cordoval) [inline cookbook/email/dev_environment.rst]: even in the dev environment

File: cookbook/email/dev_environment.rst [hunk 11]
```diff
@@ -119,6 +119,58 @@ the replaced address, so you can still see who it would have been sent to.
...
+        <swiftmailer:delivery-whitelist>/@mydomain.com$/</swiftmailer:delivery-whitelist>
+        <swiftmailer:delivery-whitelist>/^admin@specialdomain.com$/</swiftmailer:delivery-whitelist>
+    .. code-block:: php
+
+        // app/config/config_dev.php
+        $container->loadFromExtension('swiftmailer', array(
+            'delivery_address'  => "dev@example.com",
+            'delivery_whitelist' => array(
+                '/@mydomain.com$/',
+                '/^admin@specialdomain.com$/'
```
Comment (cordoval) [inline cookbook/email/dev_environment.rst]: do we use comma here? i don't know how it is in the documentation

File: cookbook/email/dev_environment.rst (hunk 6 above)
Comment (TerjeBr) [inline cookbook/email/dev_environment.rst]: I have now repeated the comments for all formats.

File: cookbook/email/dev_environment.rst (hunk 11 above)
Comment (TerjeBr) [inline cookbook/email/dev_environment.rst]: Why do you want a comma there? It is the end of the php array.

Comment (WouterJ) [inline cookbook/email/dev_environment.rst]: all array items end with a comma in Symfony CS when using a multiline array. That's to make diffs nicer when adding a new item

Comment (TerjeBr) [inline cookbook/email/dev_environment.rst]: I think that is nitpicking. The code is just an example, and may be used in code not folliwing Symfony CS. But I just changed it, none the less.

**Comment by TerjeBr** [comment]
Anything else before this can get the label finished?

File: cookbook/email/dev_environment.rst [hunk 12]
```diff
@@ -119,6 +119,66 @@ the replaced address, so you can still see who it would have been sent to.
...
+
+    .. code-block:: xml
+
+        <!-- app/config/config_dev.xml -->
+
+        <!--
+            xmlns:swiftmailer="http://symfony.com/schema/dic/swiftmailer"
+            http://symfony.com/schema/dic/swiftmailer http://symfony.com/schema/dic/swiftmailer/swiftmailer-1.0.xsd
+        -->
+
```
Comment (WouterJ) [inline cookbook/email/dev_environment.rst]: you should show a full XML config example here (I believe I've put an example in one of my comments in this PR)

File: cookbook/email/dev_environment.rst [hunk 13]
```diff
@@ -119,6 +119,66 @@ the replaced address, so you can still see who it would have been sent to.
     These are ``X-Swift-Cc`` and ``X-Swift-Bcc`` for the ``CC`` and ``BCC``
     addresses respectively.
 
+Sending to a Specified Address but with Exceptions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Suppose you want to have all email redirected to a specific address,
+(like in the above scenario to ``dev@example.com``).
+But then you may want email sent to some specific email addresses to go through
```
Comment (WouterJ) [inline cookbook/email/dev_environment.rst]: this should be joined with the previous line, only add line breaks when you're crossing the 72th character.

**Comment by WouterJ** [comment]
Sorry @TerjeBr, 2 more issues. After that, I'll flag it as finished and @weaverryan wil ltake a look at it.

**Comment by TerjeBr** [comment]
Ok. Done that.

**Comment by WouterJ** [comment]
Thanks! I've labelled it with finished now, I hope Ryan won't test your patience for too long :)

**Comment by weaverryan** [comment]
Cool feature - I did *not* know about this. I've merged this into the 2.3 branch. Thanks @TerjeBr!
